modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/distruptor/inbound/MessagePreProcessor.java

```diff
@@ -112,11 +113,19 @@ private void handleTopicRoutine(InboundEvent event, AndesMessage message, AndesC
         String messageRoutingKey = message.getMetadata().getDestination();
         //get all topic subscriptions in the cluster matching to routing key
         //including hierarchical topic case
-        List<AndesSubscription> subscriptionList;
+        Set<AndesSubscription> subscriptionList;
         try {
 
-            // TODO: This call is O(n2). In critical path. Need to improve
-            subscriptionList = subscriptionStore.getClusterSubscribersForDestination(messageRoutingKey, true);
+            // Get subscription list according to the message type
+            AndesSubscription.SubscriptionType subscriptionType;
+
+            if (MessageMetaDataType.META_DATA_MQTT == message.getMetadata().getMetaDataType()) {
+                subscriptionType = AndesSubscription.SubscriptionType.MQTT;
+            } else {
+                subscriptionType = AndesSubscription.SubscriptionType.AMQP;
+            }
```

Comment:
This was not possible since there are many metadata types for AMQP. There are three meta data types in total while there are only two subscription types.